lib/phoenix_live_view/tag_engine.ex

```diff
@@ -1160,6 +1182,76 @@ defmodule Phoenix.LiveView.TagEngine do
     end
   end
 
+  defp maybe_keyed(state, type, %{key: key_expr, for: for_expr} = tag_meta) do
+    # for now, we only support plain tags because we don't know if a function component
+    # renders to a single tag which is required by live components
+    if type != :tag do
+      raise_syntax_error!(
+        "keyed :for comprehensions only supported on regular tags, not for #{tag_meta.tag_name}",
+        tag_meta,
+        state
+      )
+    end
+
+    # we already validated that the for expression has the correct shape in
+    # validate_quoted_special_attr
+    {:<-, for_meta, [lhs, rhs]} = for_expr
+    # now we mark all eligible variables in the left-hand side as `:change_track`able
+    {lhs, variables} = mark_variables_as_change_tracked(lhs)
+    for_expr = {:<-, for_meta, [lhs, rhs]}
+
+    # now we build the new ast that we pass to the engine
+    ast =
+      quote do
+        Phoenix.LiveView.TagEngine.keyed_comprehension(
+          {unquote(state.caller.module), unquote(tag_meta.line), unquote(tag_meta.column),
+           unquote(key_expr)},
+          %{unquote_splicing(variables)},
+          do: unquote(invoke_subengine(state, :handle_end, []))
+        )
+      end
+
+    state = pop_substate_from_stack(state)
+
+    keyed_ast =
+      state
+      |> push_substate_to_stack()
+      |> update_subengine(:handle_begin, [])
+      |> update_subengine(:handle_expr, ["=", ast])
+      |> invoke_subengine(:handle_end, [])
+
+    {for_expr, keyed_ast}
+  end
+
+  defp maybe_keyed(state, _type, %{for: for_expr}) do
+    {for_expr, invoke_subengine(state, :handle_end, [])}
+  end
+
+  defp mark_variables_as_change_tracked(ast) do
+    {ast, {_, vars}} =
+      Macro.prewalk(ast, {true, []}, fn
+        # skip pinned expressions
+        {:^, _, [_expr]} = pin_expr, {true, vars} ->
+          {pin_expr, {false, vars}}
+
+        # skip the right hand side in something like <<foo::binary>>
+        {:"::", meta, [left, right]}, {true, vars} ->
```

Comment:
I believe we no longer need this now as `analyze` skips this already. It also means we solve the issue with setting the accumlator to `false` below.

I guess another option to solve above is to deal with `^` in the engine, basically we pattern match on `^var`'s AST and we handle the variable directly. Unsure which one is better.